apache_httpd-formula/apache_httpd/init.sls

```diff
@@ -0,0 +1,143 @@
+Copyright (C) 2024 Georg Pfuetzenreuter <mail+opensuse@georg-pfuetzenreuter.net>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+-#}
+
+{%- from 'apache_httpd/map.jinja' import httpd, modules, places, cmd_kwargs -%}
+{%- set sysconfig = httpd.sysconfig -%}
+{%- set mpm       = sysconfig.APACHE_MPM -%}
+
+{%- if salt['file.access']('/usr/sbin/a2enmod', 'x') %}
+{%- set have_a2enmod = True %}
+{%- set try_a2enmod = True %}
+{%- else %}
+{%- set have_a2enmod = False %}
+{#- if apache2 is not yet installed, states using a2enmod fail in test mode #}
+{%- set try_a2enmod = not opts['test'] %}
+{%- endif %}
+
+{#- no native is-active implementation in modules.systemd_service,
+    this was deemed better than reading the (potentially big) service.get_running list
+#}
+{%- if salt['cmd.retcode']('/usr/bin/systemctl is-active apache2', **cmd_kwargs) == 0 %}
+  {%- macro watch_in_restart() %}
+    - watch_in:
+        - module: apache_httpd_service_restart
+  {%- endmacro %}
+{%- else %}
+  {%- macro watch_in_restart() %}
+    {#- noop #}
+  {%- endmacro %}
+{%- endif %}
+
+{%- if httpd.purge %}
+include:
+  - .purge
+{%- endif %}
+
+apache_httpd_packages:
+  pkg.installed:
+    - pkgs:
+        - apache2-{{ mpm }}
+        {%- for module in modules %}
+          {%- if module not in httpd.internal.modules.base and module != 'cgi' %}
+        - apache2-mod_{{ module }}
+          {%- endif %}
+        {%- endfor %}
+
+apache_httpd_sysconfig:
+  suse_sysconfig.sysconfig:
+    - name: apache2
+    - key_values:
+        {%- for key, value in sysconfig.items() %}
+        {{ key }}: '"{{ value }}"'
+        {%- endfor %}
+    - require:
+        - pkg: apache_httpd_packages
+
+{%- if try_a2enmod %}
+{%- for module in modules %}
+apache_httpd_load_module-{{ module }}:
+  module.run:
+    - apache.a2enmod:
+        - mod: {{ module }}
+    - require:
+        - pkg: apache_httpd_packages
+    - unless:
+        - fun: apache.check_mod_enabled
+          mod: {{ module }}
+    {{ watch_in_restart() }}
+{%- endfor %} {#- close configured modules loop #}
+{%- endif %} {#- close a2enmod check #}
+
+{%- if have_a2enmod %}
+{%- for module in salt['cmd.run_stdout']('/usr/sbin/a2enmod -l', **cmd_kwargs).split() %}
+  {%- if module not in modules %}
+apache_httpd_unload_module-{{ module }}:
+  module.run:
+    - apache.a2dismod:
+        - mod: {{ module }}
```

Comment:
So you first call `a2enmod` for the requested modules, and then `a2dismod` everything else.

Wouldn't it be easier to just write the list of requested modules to the sysconfig file? (AFAIK that's what a2enmod and a2dismod do.)

Reply:
That is indeed what a2{en,dis}mod do, but to manage `APACHE_MODULES` directly I would need a separate sysconfig handler as to not hook changes to the other sysconfig settings to a full service restart, hence I think it wouldn't reduce the complexity much.
Also I get to use the native functionality.